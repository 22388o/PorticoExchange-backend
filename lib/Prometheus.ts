import express, { Response } from 'express';
import { collectDefaultMetrics, Gauge, Registry } from 'prom-client';
import Logger from './Logger';
import StatsRepository from './db/repositories/StatsRepository';

type PrometheusConfig = {
  host?: string;
  port?: number;
};

class Prometheus {
  private static readonly metric_prefix = 'boltz_';

  private readonly nodeRegistry?: Registry;
  private readonly swapRegistry?: Registry;

  constructor(
    private readonly logger: Logger,
    private readonly config?: PrometheusConfig,
  ) {
    if (
      this.config === undefined ||
      Object.values(this.config).some((value) => value === undefined)
    ) {
      this.logger.warn(
        'Disabling metrics server because of missing configuration',
      );
      return;
    }

    this.nodeRegistry = new Registry();
    this.swapRegistry = new Registry();
    this.registerMetrics();
  }

  public start = async (): Promise<void> => {
    if (
      [this.nodeRegistry, this.swapRegistry].some(
        (registry) => registry === undefined,
      )
    ) {
      return;
    }

    const respondMetrics = (registry: Registry) => {
      return async (_: unknown, res: Response) => {
        res.setHeader('Content-Type', registry.contentType);
        res.send(await registry.metrics());
      };
    };

    const app = express();

    app.route('/metrics').get(respondMetrics(this.nodeRegistry!));
    app.route('/swapMetrics').get(respondMetrics(this.swapRegistry!));

    new Promise<void>((resolve) => {
      app.listen(this.config!.port!, this.config!.host!, () => {
        this.logger.info(
          `Metrics server listening on: ${this.config!.host}:${
            this.config!.port
          }`,
        );
        resolve();
      });
    });
  };

  private registerMetrics = () => {
    collectDefaultMetrics({
      register: this.nodeRegistry,
    });

    this.swapRegistry!.registerMetric(
      new Gauge({
        name: `${Prometheus.metric_prefix}swap_counts`,
        labelNames: ['pair', 'type', 'status'],
        help: 'number of swaps',
        collect: async function () {
          const counts = await StatsRepository.getSwapCounts();

          counts.forEach((count) =>
            this.set(
              { pair: count.pair, type: count.type, status: count.status },
              count.count,
            ),
          );
        },
      }),
    );

    this.swapRegistry!.registerMetric(
      new Gauge({
        name: `${Prometheus.metric_prefix}swap_volume`,
        labelNames: ['pair', 'type'],
        help: 'volume of swaps',
        collect: async function () {
          const counts = await StatsRepository.getVolumePerPairType();

          counts.forEach((volume) =>
            this.set({ pair: volume.pair, type: volume.type }, volume.volume),
          );
        },
      }),
    );

    this.swapRegistry!.registerMetric(
      new Gauge({
        name: `${Prometheus.metric_prefix}swap_pending_counts`,
        labelNames: ['pair', 'type'],
        help: 'count of pending swaps',
        collect: async function () {
          const counts = await StatsRepository.getPendingSwapsCounts();

          counts.forEach((volume) =>
            this.set({ pair: volume.pair, type: volume.type }, volume.count),
          );
        },
      }),
    );

    this.swapRegistry!.registerMetric(
      new Gauge({
        name: `${Prometheus.metric_prefix}swap_locked_funds`,
        labelNames: ['pair', 'type'],
        help: 'coins locked in pending swaps',
        collect: async function () {
          const lockedFunds = await StatsRepository.getLockedFunds();

          lockedFunds.forEach((locked) =>
            this.set({ pair: locked.pair, type: 'reverse' }, locked.locked),
          );
        },
      }),
    );
  };
}

export default Prometheus;
export { PrometheusConfig };
