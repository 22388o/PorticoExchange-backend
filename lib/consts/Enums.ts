export enum ErrorCodePrefix {
  General = 0,
  Grpc = 1,
  Service = 2,
  Wallet = 3,
  Chain = 4,
  Lnd = 5,
  Swap = 6,
  Rates = 7,
  Bakcup = 8,
}

export enum ClientStatus {
  Disconnected,
  Connected,
  OutOfSync,
}

export enum OrderSide {
  BUY,
  SELL,
}

export enum SwapUpdateEvent {
  InvoicePaid = 'invoice.paid',
  InvoiceSettled = 'invoice.settled',
  InvoiceFailedToPay = 'invoice.failedToPay',

  // Means that Boltz is waiting for the user to send a transaction
  TransactionWaiting = 'transaction.waiting',

  TransactionMempool = 'transaction.mempool',
  TransactionClaimed = 'transaction.claimed',
  TransactionConfirmed = 'transaction.confirmed',

  // Transaction events with the prefix "boltz" are related to transactions sent by Boltz
  BoltzTransactionMempool = 'boltz.transaction.mempool',
  BoltzTransactionConfirmed = 'boltz.transaction.confirmed',
  BoltzTransactionRefunded = 'boltz.transaction.refunded',

  SwapAborted = 'swap.aborted',
  SwapExpired = 'swap.expired',
}

export enum ServiceWarning {
  ReverseSwapsDisabled = 'reverse.swaps.disabled',
}

export enum Network {
  Mainnet = 'mainnet',
  Testnet = 'testnet',
  Simnet = 'simnet',
  Regtest = 'regtest',
}

export enum SwapType {
  Submarine = 'submarine',
  ReverseSubmarine = 'reversesubmarine',

  ChainToChain = 'chaintochain',
}
