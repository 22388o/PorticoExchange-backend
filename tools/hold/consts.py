PLUGIN_NAME = "hold"

TIMEOUT_CANCEL = 60
TIMEOUT_CHECK_INTERVAL = 10
