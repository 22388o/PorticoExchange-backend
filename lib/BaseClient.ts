import { EventEmitter } from 'events';
import Logger from './Logger';
import { racePromise } from './PromiseUtils';
import { ClientStatus } from './consts/Enums';

interface IBaseClient {
  on(event: 'status.changed', listener: (status: ClientStatus) => void): void;
  emit(event: 'status.changed', status: ClientStatus): void;
}

class BaseClient extends EventEmitter implements IBaseClient {
  private status = ClientStatus.Disconnected;

  protected readonly RECONNECT_INTERVAL = 5000;
  protected reconnectionTimer?: any;

  constructor(
    protected readonly logger: Logger,
    private readonly name: string,
  ) {
    super();
  }

  public isConnected(): boolean {
    return this.status === ClientStatus.Connected;
  }

  public isDisconnected(): boolean {
    return this.status === ClientStatus.Disconnected;
  }

  public isOutOfSync(): boolean {
    return this.status === ClientStatus.OutOfSync;
  }

  public setClientStatus = (status: ClientStatus): void => {
    if (this.status === status) {
      return;
    }

    this.status = status;

    (status === ClientStatus.Connected ? this.logger.info : this.logger.error)(
      `${this.name} status changed: ${status}`,
    );

    this.emit('status.changed', status);
  };

  public raceCall = <T>(
    promise: (() => Promise<T>) | Promise<T>,
    raceHandler: (reason?: any) => void,
    raceTimeout: number,
  ): Promise<T> => {
    return racePromise(
      promise,
      (reject) => {
        this.setClientStatus(ClientStatus.Disconnected);
        raceHandler(reject);
      },
      raceTimeout,
    );
  };

  protected clearReconnectTimer = (): void => {
    if (this.reconnectionTimer) {
      clearInterval(this.reconnectionTimer);
      this.reconnectionTimer = undefined;
    }
  };
}

export default BaseClient;
export { IBaseClient };
