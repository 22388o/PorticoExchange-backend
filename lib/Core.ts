import zkp from '@vulpemventures/secp256k1-zkp';
import { Network as NetworkLiquid } from 'liquidjs-lib/src/networks';
import { address, Network, Transaction, TxOutput } from 'bitcoinjs-lib';
import {
  address as addressLiquid,
  confidential,
  Transaction as TransactionLiquid,
  TxOutput as TxOutputLiquid,
} from 'liquidjs-lib';
import {
  ClaimDetails as ClaimDetailsBitcoin,
  RefundDetails as RefundDetailsBitcoin,
} from 'boltz-core/dist/lib/consts/Types';
import {
  ClaimDetails as ClaimDetailsLiquid,
  RefundDetails as RefundDetailsLiquid,
} from 'boltz-core-liquid-michael1011/dist/consts/Types';
import {
  constructClaimTransaction as constructClaimTransactionBitcoin,
  constructRefundTransaction as constructRefundTransactionBitcoin,
  detectPreimage as detectPreimageBitcoin,
  detectSwap as detectSwapBitcoin,
} from 'boltz-core';
import {
  constructClaimTransaction as constructClaimTransactionLiquid,
  constructRefundTransaction as constructRefundTransactionLiquid,
  detectPreimage as detectPreimageLiquid,
  detectSwap as detectSwapLiquid,
  targetFee,
  prepareConfidential,
} from 'boltz-core-liquid-michael1011';
import Wallet from './wallet/Wallet';
import ChainClient from './chain/ChainClient';
import { CurrencyType } from './consts/Enums';
import WalletLiquid from './wallet/WalletLiquid';
import {
  getHexString,
  reverseBuffer,
  calculateUtxoTransactionFee,
  calculateLiquidTransactionFee,
} from './Utils';

let confi: confidential.Confidential;

export const setup = async () => {
  const zkpLib = await zkp();
  confi = new confidential.Confidential(zkpLib);
  prepareConfidential(zkpLib);
};

export const parseTransaction = (
  type: CurrencyType,
  rawTx: string | Buffer,
): Transaction | TransactionLiquid => {
  if (rawTx instanceof Buffer) {
    return isBitcoin(type)
      ? Transaction.fromBuffer(rawTx)
      : TransactionLiquid.fromBuffer(rawTx);
  } else {
    return isBitcoin(type)
      ? Transaction.fromHex(rawTx)
      : TransactionLiquid.fromHex(rawTx);
  }
};

export const fromOutputScript = (
  type: CurrencyType,
  outputScript: Buffer,
  network: Network | NetworkLiquid,
) => {
  return isBitcoin(type)
    ? address.fromOutputScript(outputScript, network)
    : addressLiquid.fromOutputScript(outputScript, network as NetworkLiquid);
};

export const toOutputScript = (
  type: CurrencyType,
  toDecode: string,
  network: Network | NetworkLiquid,
) => {
  return isBitcoin(type)
    ? address.toOutputScript(toDecode, network)
    : addressLiquid.toOutputScript(toDecode, network as NetworkLiquid);
};

export const getOutputValue = (
  wallet: Wallet,
  output: TxOutput | TxOutputLiquid,
) => {
  if (isBitcoin(wallet.type)) {
    return output.value as number;
  }

  let value: number;
  let asset: Buffer;

  if ((output as TxOutputLiquid).rangeProof?.length !== 0) {
    const unblinded = confi.unblindOutputWithKey(
      output as TxOutputLiquid,
      (wallet as WalletLiquid).deriveBlindingKeyFromScript(output.script)
        .privateKey!,
    );

    value = Number(unblinded.value);
    asset = unblinded.asset;
  } else {
    value = confidential.confidentialValueToSatoshi(output.value as Buffer);
    asset = (output as TxOutputLiquid).asset;
    // Remove the un-confidential prefix
    asset = asset.slice(1, asset.length);
  }

  if (
    getHexString(reverseBuffer(asset)) !==
    (wallet.network as NetworkLiquid).assetHash
  ) {
    return 0;
  }

  return value;
};

export const detectSwap = (
  type: CurrencyType,
  redeemScript: Buffer,
  transaction: Transaction | TransactionLiquid,
) => {
  return isBitcoin(type)
    ? detectSwapBitcoin(redeemScript, transaction as Transaction)
    : detectSwapLiquid(redeemScript, transaction as TransactionLiquid);
};

export const detectPreimage = (
  type: CurrencyType,
  vin: number,
  transaction: Transaction | TransactionLiquid,
) => {
  return isBitcoin(type)
    ? detectPreimageBitcoin(vin, transaction as Transaction)
    : detectPreimageLiquid(vin, transaction as TransactionLiquid);
};

export const constructClaimTransaction = (
  wallet: Wallet,
  claimDetails: ClaimDetailsBitcoin[] | ClaimDetailsLiquid[],
  destinationAddress: string,
  feePerVbyte: number,
  assetHash?: string,
) => {
  if (isBitcoin(wallet.type)) {
    return constructClaimTransactionBitcoin(
      claimDetails as ClaimDetailsBitcoin[],
      wallet.decodeAddress(destinationAddress),
      feePerVbyte,
      true,
    );
  }

  const decodedAddress = addressLiquid.fromConfidential(destinationAddress);
  return targetFee(feePerVbyte, (fee) =>
    constructClaimTransactionLiquid(
      populateBlindingKeys(
        wallet as WalletLiquid,
        claimDetails as ClaimDetailsLiquid[],
      ),
      decodedAddress.scriptPubKey!,
      fee,
      true,
      assetHash,
      decodedAddress.blindingKey,
    ),
  );
};

export const constructRefundTransaction = (
  wallet: Wallet,
  refundDetails: RefundDetailsBitcoin[] | RefundDetailsLiquid[],
  destinationAddress: string,
  timeoutBlockHeight: number,
  feePerVbyte: number,
  assetHash?: string,
) => {
  if (isBitcoin(wallet.type)) {
    return constructRefundTransactionBitcoin(
      refundDetails as RefundDetailsBitcoin[],
      wallet.decodeAddress(destinationAddress),
      timeoutBlockHeight,
      feePerVbyte,
      true,
    );
  }

  const decodedAddress = addressLiquid.fromConfidential(destinationAddress);
  return targetFee(feePerVbyte, (fee) =>
    constructRefundTransactionLiquid(
      populateBlindingKeys(
        wallet as WalletLiquid,
        refundDetails as RefundDetailsLiquid[],
      ),
      decodedAddress.scriptPubKey!,
      timeoutBlockHeight,
      fee,
      true,
      assetHash,
      decodedAddress.blindingKey,
    ),
  );
};

export const calculateTransactionFee = async (
  chainClient: ChainClient,
  transaction: Transaction | TransactionLiquid,
) => {
  return transaction instanceof Transaction
    ? calculateUtxoTransactionFee(chainClient, transaction as Transaction)
    : calculateLiquidTransactionFee(transaction as TransactionLiquid);
};

export const getAssetHash = (
  type: CurrencyType,
  network: Network,
): string | undefined => {
  return isBitcoin(type) ? undefined : (network as NetworkLiquid).assetHash;
};

const isBitcoin = (type: CurrencyType) => type === CurrencyType.BitcoinLike;

const populateBlindingKeys = <
  T extends ClaimDetailsLiquid | RefundDetailsLiquid,
>(
  wallet: WalletLiquid,
  utxos: T[],
): T[] => {
  for (const utxo of utxos) {
    utxo.blindingPrivKey = wallet.deriveBlindingKeyFromScript(
      utxo.script,
    ).privateKey!;
  }

  return utxos;
};

export {
  ClaimDetailsBitcoin,
  ClaimDetailsLiquid,
  RefundDetailsBitcoin,
  RefundDetailsLiquid,
};
