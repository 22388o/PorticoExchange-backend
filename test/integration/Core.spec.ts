import { BIP32Factory } from 'bip32';
import { randomBytes } from 'crypto';
import * as ecc from 'tiny-secp256k1';
import { SLIP77Factory } from 'slip77';
import { Networks, swapTree } from 'boltz-core';
import { toXOnly } from 'bitcoinjs-lib/src/psbt/bip371';
import { generateMnemonic, mnemonicToSeedSync } from 'bip39';
import { networks, Transaction as TransactionLiquid } from 'liquidjs-lib';
import Logger from '../../lib/Logger';
import Database from '../../lib/db/Database';
import Wallet from '../../lib/wallet/Wallet';
import { ECPair } from '../../lib/ECPairHelper';
import { CurrencyType } from '../../lib/consts/Enums';
import { bitcoinClient, elementsClient } from './Nodes';
import WalletLiquid from '../../lib/wallet/WalletLiquid';
import CoreWalletProvider from '../../lib/wallet/providers/CoreWalletProvider';
import ElementsWalletProvider from '../../lib/wallet/providers/ElementsWalletProvider';
import {
  extractRefundPublicKeyFromSwapTree,
  fromOutputScript,
  getOutputValue,
  setup,
  toOutputScript,
} from '../../lib/Core';

describe('Core', () => {
  const bip32 = BIP32Factory(ecc);
  const slip77 = SLIP77Factory(ecc);

  const database = new Database(Logger.disabledLogger, ':memory:');

  let wallet: Wallet;
  let walletLiquid: WalletLiquid;

  beforeAll(async () => {
    await Promise.all([setup(), database.init()]);

    const initWallet = (w: Wallet, network: any) => {
      w.initKeyProvider(
        network,
        'm/0/0',
        0,
        bip32.fromSeed(mnemonicToSeedSync(generateMnemonic())),
      );
    };

    wallet = new Wallet(
      Logger.disabledLogger,
      CurrencyType.BitcoinLike,
      new CoreWalletProvider(Logger.disabledLogger, bitcoinClient),
    );
    initWallet(wallet, Networks.bitcoinRegtest);

    walletLiquid = new WalletLiquid(
      Logger.disabledLogger,
      new ElementsWalletProvider(Logger.disabledLogger, elementsClient),
      slip77.fromSeed(generateMnemonic()),
    );
    initWallet(walletLiquid, networks.regtest);

    await bitcoinClient.connect();
    await elementsClient.connect();

    // To have a blinded output in the Elements wallet
    await elementsClient.sendToAddress(
      await elementsClient.getNewAddress(),
      10 ** 8,
    );

    await Promise.all([bitcoinClient.generate(1), elementsClient.generate(1)]);
  });

  afterAll(async () => {
    await Promise.all([bitcoinClient.generate(1), elementsClient.generate(1)]);

    bitcoinClient.disconnect();
    elementsClient.disconnect();
  });

  test('should get output value of Bitcoin transactions', async () => {
    const outputAmount = 420234;
    const { transaction, vout } = await wallet.sendToAddress(
      await wallet.getAddress(),
      outputAmount,
    );

    expect(getOutputValue(wallet, transaction!.outs[vout!])).toEqual(
      outputAmount,
    );
  });

  test('should get output value of unblinded Liquid transactions', async () => {
    const outputAmount = 562312;
    const { transaction, vout } = await walletLiquid.sendToAddress(
      fromOutputScript(
        CurrencyType.Liquid,
        toOutputScript(
          CurrencyType.Liquid,
          await walletLiquid.getAddress(),
          networks.regtest,
        ),
        networks.regtest,
      ),
      outputAmount,
    );

    expect(
      getOutputValue(
        walletLiquid,
        (transaction as TransactionLiquid).outs[vout!],
      ),
    ).toEqual(outputAmount);

    // Wrong asset
    walletLiquid['network'] = networks.liquid;
    expect(
      getOutputValue(
        walletLiquid,
        (transaction as TransactionLiquid).outs[vout!],
      ),
    ).toEqual(0);
    walletLiquid['network'] = networks.regtest;
  });

  test('should get output value of blinded Liquid transactions', async () => {
    const script = toOutputScript(
      CurrencyType.Liquid,
      await walletLiquid.getAddress(),
      walletLiquid.network,
    );

    const outputAmount = 1245412;
    const { transaction, vout } = await walletLiquid.sendToAddress(
      walletLiquid.encodeAddress(script),
      outputAmount,
    );

    expect(
      getOutputValue(
        walletLiquid,
        (transaction as TransactionLiquid).outs[vout!],
      ),
    ).toEqual(outputAmount);

    // Wrong asset
    walletLiquid['network'] = networks.liquid;
    expect(
      getOutputValue(
        walletLiquid,
        (transaction as TransactionLiquid).outs[vout!],
      ),
    ).toEqual(0);
    walletLiquid['network'] = networks.regtest;
  });

  test('should extract refund public key from swap tree', () => {
    const refundKey = ECPair.makeRandom().publicKey;
    const tree = swapTree(
      false,
      randomBytes(32),
      ECPair.makeRandom().publicKey,
      refundKey,
      1,
    );

    expect(extractRefundPublicKeyFromSwapTree(tree)).toEqual(
      toXOnly(refundKey),
    );
  });
});
