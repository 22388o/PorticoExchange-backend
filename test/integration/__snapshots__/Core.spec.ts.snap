// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core should create Musig 1`] = `
{
  "data": [
    16,
    131,
    62,
    35,
    164,
    197,
    26,
    253,
    60,
    182,
    108,
    147,
    52,
    59,
    37,
    170,
    124,
    172,
    207,
    152,
    150,
    225,
    133,
    71,
    250,
    117,
    188,
    170,
    34,
    24,
    180,
    26,
  ],
  "type": "Buffer",
}
`;

exports[`Core should tweak Musig (liquid: false) 1`] = `
{
  "data": [
    71,
    125,
    26,
    124,
    73,
    150,
    189,
    182,
    94,
    245,
    120,
    137,
    63,
    61,
    214,
    121,
    170,
    149,
    106,
    237,
    133,
    169,
    224,
    96,
    124,
    56,
    255,
    108,
    42,
    129,
    57,
    40,
  ],
  "type": "Buffer",
}
`;

exports[`Core should tweak Musig (liquid: true) 1`] = `
{
  "data": [
    63,
    197,
    137,
    12,
    129,
    106,
    176,
    248,
    79,
    212,
    79,
    137,
    115,
    63,
    170,
    0,
    233,
    101,
    213,
    254,
    85,
    32,
    130,
    88,
    171,
    99,
    212,
    235,
    194,
    6,
    131,
    151,
  ],
  "type": "Buffer",
}
`;
